Portfolio Valuation

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
# Set the display options to show all columns
pd.set_option('display.max_columns', None)

In [3]:
data = pd.read_csv(r'C:\Users\phili\Dropbox\Documents\2024-data science\port_val.csv')
data.head()

Column1  Origination Amount  31.05.2019  30.06.2019  31.07.2019  \
0  31.05.2019         10018746.17  1443069.08  3332200.33  1328138.75   
1  30.06.2019         10868379.04        0.00  1392751.60  3011884.91   
2  31.07.2019         10733932.61        0.00        0.00  1537650.24   
3  31.08.2019         12558727.02        0.00        0.00        0.00   
4  30.09.2019         14505071.44        0.00        0.00        0.00   

   31.08.2019  30.09.2019  31.10.2019  30.11.2019  31.12.2019  31.01.2020  \
0   928085.74   736418.27   539403.31   427557.86   324459.32   237056.39   
1  1237868.70   970929.28   892351.83   668767.02   505612.59   419598.74   
2  2953335.55  1208316.08   879375.19   711016.84   658251.40   503465.03   
3  1617681.94  4082016.00  1387474.94  1247623.59   886293.35   694348.63   
4        0.00  1992242.84  3930445.60  1394620.78  1227905.58   939424.54   

   29.02.2020  31.03.2020  30.04.2020  31.05.2020  30.06.2020  31.07.2020  \
0   168364.60   116684.68    92699.67    63399.66    53265.12    37121.13   
1   329262.69   255222.42   198833.96   161996.73   138461.91    92346.68   
2   423045.23   302575.54   258652.52   191798.05   170027.54   127574.33   
3   571024.44   417223.56   336686.08   253556.20   200066.59   151859.74   
4   802871.19   628429.48   589692.85   457299.31   323764.87   288152.28   

   31.08.2020  30.09.2020  31.10.2020  30.11.2020  31.12.2020  
0    29787.10    24524.90    18085.94    16581.01    11442.97  
1    79641.30    63457.44    52373.85    43374.70    37404.87  
2   110301.21    89766.69    64746.84    61408.92    50312.70  
3   109973.00    90228.14    70661.50    53102.83    47069.84  
4   239872.99   192246.98   171550.69   142575.97   116853.05

In [4]:
df = data.copy()

In [5]:
# Convert column headers to real dates
df.columns = ['Column1', 'Origination Amount'] + [pd.to_datetime(col, format='%d.%m.%Y') for col in df.columns[2:]]
# Rename columns into "Repayment (I) + date" format
df.columns = df.columns[:2].tolist() + [f'Repayment ({i}) {col.strftime("%d.%m.%Y")}' for i, col in enumerate(df.columns[2:], start=1)]

df.head()

Column1  Origination Amount  Repayment (1) 31.05.2019  \
0  31.05.2019         10018746.17                1443069.08   
1  30.06.2019         10868379.04                      0.00   
2  31.07.2019         10733932.61                      0.00   
3  31.08.2019         12558727.02                      0.00   
4  30.09.2019         14505071.44                      0.00   

   Repayment (2) 30.06.2019  Repayment (3) 31.07.2019  \
0                3332200.33                1328138.75   
1                1392751.60                3011884.91   
2                      0.00                1537650.24   
3                      0.00                      0.00   
4                      0.00                      0.00   

   Repayment (4) 31.08.2019  Repayment (5) 30.09.2019  \
0                 928085.74                 736418.27   
1                1237868.70                 970929.28   
2                2953335.55                1208316.08   
3                1617681.94                4082016.00   
4                      0.00                1992242.84   

   Repayment (6) 31.10.2019  Repayment (7) 30.11.2019  \
0                 539403.31                 427557.86   
1                 892351.83                 668767.02   
2                 879375.19                 711016.84   
3                1387474.94                1247623.59   
4                3930445.60                1394620.78   

   Repayment (8) 31.12.2019  Repayment (9) 31.01.2020  \
0                 324459.32                 237056.39   
1                 505612.59                 419598.74   
2                 658251.40                 503465.03   
3                 886293.35                 694348.63   
4                1227905.58                 939424.54   

   Repayment (10) 29.02.2020  Repayment (11) 31.03.2020  \
0                  168364.60                  116684.68   
1                  329262.69                  255222.42   
2                  423045.23                  302575.54   
3                  571024.44                  417223.56   
4                  802871.19                  628429.48   

   Repayment (12) 30.04.2020  Repayment (13) 31.05.2020  \
0                   92699.67                   63399.66   
1                  198833.96                  161996.73   
2                  258652.52                  191798.05   
3                  336686.08                  253556.20   
4                  589692.85                  457299.31   

   Repayment (14) 30.06.2020  Repayment (15) 31.07.2020  \
0                   53265.12                   37121.13   
1                  138461.91                   92346.68   
2                  170027.54                  127574.33   
3                  200066.59                  151859.74   
4                  323764.87                  288152.28   

   Repayment (16) 31.08.2020  Repayment (17) 30.09.2020  \
0                   29787.10                   24524.90   
1                   79641.30                   63457.44   
2                  110301.21                   89766.69   
3                  109973.00                   90228.14   
4                  239872.99                  192246.98   

   Repayment (18) 31.10.2020  Repayment (19) 30.11.2020  \
0                   18085.94                   16581.01   
1                   52373.85                   43374.70   
2                   64746.84                   61408.92   
3                   70661.50                   53102.83   
4                  171550.69                  142575.97   

   Repayment (20) 31.12.2020  
0                   11442.97  
1                   37404.87  
2                   50312.70  
3                   47069.84  
4                  116853.05

In [6]:
df.shape

(20, 22)

In [7]:
hist_repay_perc = df.copy()
hist_repay_perc['Column1'] = pd.to_datetime(df['Column1'], dayfirst=True)

hist_repay_perc.head()

Column1  Origination Amount  Repayment (1) 31.05.2019  \
0 2019-05-31         10018746.17                1443069.08   
1 2019-06-30         10868379.04                      0.00   
2 2019-07-31         10733932.61                      0.00   
3 2019-08-31         12558727.02                      0.00   
4 2019-09-30         14505071.44                      0.00   

   Repayment (2) 30.06.2019  Repayment (3) 31.07.2019  \
0                3332200.33                1328138.75   
1                1392751.60                3011884.91   
2                      0.00                1537650.24   
3                      0.00                      0.00   
4                      0.00                      0.00   

   Repayment (4) 31.08.2019  Repayment (5) 30.09.2019  \
0                 928085.74                 736418.27   
1                1237868.70                 970929.28   
2                2953335.55                1208316.08   
3                1617681.94                4082016.00   
4                      0.00                1992242.84   

   Repayment (6) 31.10.2019  Repayment (7) 30.11.2019  \
0                 539403.31                 427557.86   
1                 892351.83                 668767.02   
2                 879375.19                 711016.84   
3                1387474.94                1247623.59   
4                3930445.60                1394620.78   

   Repayment (8) 31.12.2019  Repayment (9) 31.01.2020  \
0                 324459.32                 237056.39   
1                 505612.59                 419598.74   
2                 658251.40                 503465.03   
3                 886293.35                 694348.63   
4                1227905.58                 939424.54   

   Repayment (10) 29.02.2020  Repayment (11) 31.03.2020  \
0                  168364.60                  116684.68   
1                  329262.69                  255222.42   
2                  423045.23                  302575.54   
3                  571024.44                  417223.56   
4                  802871.19                  628429.48   

   Repayment (12) 30.04.2020  Repayment (13) 31.05.2020  \
0                   92699.67                   63399.66   
1                  198833.96                  161996.73   
2                  258652.52                  191798.05   
3                  336686.08                  253556.20   
4                  589692.85                  457299.31   

   Repayment (14) 30.06.2020  Repayment (15) 31.07.2020  \
0                   53265.12                   37121.13   
1                  138461.91                   92346.68   
2                  170027.54                  127574.33   
3                  200066.59                  151859.74   
4                  323764.87                  288152.28   

   Repayment (16) 31.08.2020  Repayment (17) 30.09.2020  \
0                   29787.10                   24524.90   
1                   79641.30                   63457.44   
2                  110301.21                   89766.69   
3                  109973.00                   90228.14   
4                  239872.99                  192246.98   

   Repayment (18) 31.10.2020  Repayment (19) 30.11.2020  \
0                   18085.94                   16581.01   
1                   52373.85                   43374.70   
2                   64746.84                   61408.92   
3                   70661.50                   53102.83   
4                  171550.69                  142575.97   

   Repayment (20) 31.12.2020  
0                   11442.97  
1                   37404.87  
2                   50312.70  
3                   47069.84  
4                  116853.05

Based on the provided data, compute the historical repayment percentages, i.e. every repayment’s share of the origination amount

In [8]:
dff = hist_repay_perc.copy()

# Calculate historical repayment percentages
repayment_cols = dff.columns[2:]
for col in repayment_cols:
    dff[col + '_Repay_perc'] = round((dff[col] / dff['Origination Amount']) * 100, 2)

# Display the DataFrame with historical repayment percentages
historical_repayment_percentages = dff.iloc[:, 2:].copy()

# Delete columns that don't contain "Repay_perc"
columns_to_keep = [col for col in historical_repayment_percentages.columns if 'Repay_perc' in col]
historical_repayment_percentages = historical_repayment_percentages[columns_to_keep]

# Extract "Column1" and "Origination Amount" columns from dff
columns_to_add = ['Column1', 'Origination Amount']
additional_columns = dff[columns_to_add]

# Concatenate additional columns with historical_repayment_percentages
historical_repayment_percentages = pd.concat([additional_columns, historical_repayment_percentages], axis=1)

historical_repayment_percentages.columns = historical_repayment_percentages.columns.str.replace('_Repay_perc', '')
historical_repayment_percentages.columns = historical_repayment_percentages.columns.str.replace('Repayment', 'Historical_repay_percent')

historical_repayment_percentages.head()


Column1  Origination Amount  Historical_repay_percent (1) 31.05.2019  \
0 2019-05-31         10018746.17                                     14.4   
1 2019-06-30         10868379.04                                      0.0   
2 2019-07-31         10733932.61                                      0.0   
3 2019-08-31         12558727.02                                      0.0   
4 2019-09-30         14505071.44                                      0.0   

   Historical_repay_percent (2) 30.06.2019  \
0                                    33.26   
1                                    12.81   
2                                     0.00   
3                                     0.00   
4                                     0.00   

   Historical_repay_percent (3) 31.07.2019  \
0                                    13.26   
1                                    27.71   
2                                    14.33   
3                                     0.00   
4                                     0.00   

   Historical_repay_percent (4) 31.08.2019  \
0                                     9.26   
1                                    11.39   
2                                    27.51   
3                                    12.88   
4                                     0.00   

   Historical_repay_percent (5) 30.09.2019  \
0                                     7.35   
1                                     8.93   
2                                    11.26   
3                                    32.50   
4                                    13.73   

   Historical_repay_percent (6) 31.10.2019  \
0                                     5.38   
1                                     8.21   
2                                     8.19   
3                                    11.05   
4                                    27.10   

   Historical_repay_percent (7) 30.11.2019  \
0                                     4.27   
1                                     6.15   
2                                     6.62   
3                                     9.93   
4                                     9.61   

   Historical_repay_percent (8) 31.12.2019  \
0                                     3.24   
1                                     4.65   
2                                     6.13   
3                                     7.06   
4                                     8.47   

   Historical_repay_percent (9) 31.01.2020  \
0                                     2.37   
1                                     3.86   
2                                     4.69   
3                                     5.53   
4                                     6.48   

   Historical_repay_percent (10) 29.02.2020  \
0                                      1.68   
1                                      3.03   
2                                      3.94   
3                                      4.55   
4                                      5.54   

   Historical_repay_percent (11) 31.03.2020  \
0                                      1.16   
1                                      2.35   
2                                      2.82   
3                                      3.32   
4                                      4.33   

   Historical_repay_percent (12) 30.04.2020  \
0                                      0.93   
1                                      1.83   
2                                      2.41   
3                                      2.68   
4                                      4.07   

   Historical_repay_percent (13) 31.05.2020  \
0                                      0.63   
1                                      1.49   
2                                      1.79   
3                                      2.02   
4                                      3.15   

   Historical_repay_percent (14) 30.06.2020  \
0                                      0.53   
1                                      1.27   
2                                      1.58   
3     

Compute the expected repayment percentages for all vintages over the lifetime of the loans. Details on how the expected repayment percentages are to be computed can be found in the attached assumptions PDF below.

In [9]:
# Initialize a list to store expected repayment percentages for each vintage
expected_repayment_percentages = []

# Historical repayment percentages columns
repayment_cols = historical_repayment_percentages.columns[2:]

# Iterate over each vintage
for index, row in historical_repayment_percentages.iterrows():
    vintage = row['Column1']
    origination_amount = row['Origination Amount']
    expected_repayments = []
    
    # Iterate over each repayment period for the vintage
    for i, col in enumerate(repayment_cols, start=1):
        repayment_percentage = row[col]
        
        # Calculate expected repayment percentage based on the formula
        if i == 1:
            expected_repayment = repayment_percentage
        elif i == 2:
            if index == 4: # Check if vintage is December 2020
                expected_repayment = 2 * expected_repayments[0]
            else:
                expected_repayment = expected_repayments[0]
        else:
            expected_repayment = repayment_percentage

        expected_repayments.append(expected_repayment)
    
    expected_repayment_percentages.append(expected_repayments)

# Convert the list of expected repayment percentages into a DataFrame
expected_repayments_df = pd.DataFrame(expected_repayment_percentages, columns=repayment_cols)

# Insert 'Column1' and 'Origination Amount' columns at the beginning of the DataFrame
expected_repayments_df.insert(0, 'Column1', historical_repayment_percentages['Column1'])
expected_repayments_df.insert(1, 'Origination Amount', historical_repayment_percentages['Origination Amount'])

# Rename columns with 'expected_repay'
expected_repayments_df.columns = expected_repayments_df.columns.str.replace('Historical_repay_percent', 'expected_repayment_perc')
expected_repayments_df.columns = expected_repayments_df.columns.str.replace('_Repay_perc', '')

expected_repayments_df.head()


Column1  Origination Amount  expected_repayment_perc (1) 31.05.2019  \
0 2019-05-31         10018746.17                                    14.4   
1 2019-06-30         10868379.04                                     0.0   
2 2019-07-31         10733932.61                                     0.0   
3 2019-08-31         12558727.02                                     0.0   
4 2019-09-30         14505071.44                                     0.0   

   expected_repayment_perc (2) 30.06.2019  \
0                                    14.4   
1                                     0.0   
2                                     0.0   
3                                     0.0   
4                                     0.0   

   expected_repayment_perc (3) 31.07.2019  \
0                                   13.26   
1                                   27.71   
2                                   14.33   
3                                    0.00   
4                                    0.00   

   expected_repayment_perc (4) 31.08.2019  \
0                                    9.26   
1                                   11.39   
2                                   27.51   
3                                   12.88   
4                                    0.00   

   expected_repayment_perc (5) 30.09.2019  \
0                                    7.35   
1                                    8.93   
2                                   11.26   
3                                   32.50   
4                                   13.73   

   expected_repayment_perc (6) 31.10.2019  \
0                                    5.38   
1                                    8.21   
2                                    8.19   
3                                   11.05   
4                                   27.10   

   expected_repayment_perc (7) 30.11.2019  \
0                                    4.27   
1                                    6.15   
2                                    6.62   
3                                    9.93   
4                                    9.61   

   expected_repayment_perc (8) 31.12.2019  \
0                                    3.24   
1                                    4.65   
2                                    6.13   
3                                    7.06   
4                                    8.47   

   expected_repayment_perc (9) 31.01.2020  \
0                                    2.37   
1                                    3.86   
2                                    4.69   
3                                    5.53   
4                                    6.48   

   expected_repayment_perc (10) 29.02.2020  \
0                                     1.68   
1                                     3.03   
2                                     3.94   
3                                     4.55   
4                                     5.54   

   expected_repayment_perc (11) 31.03.2020  \
0                                     1.16   
1                                     2.35   
2                                     2.82   
3                                     3.32   
4                                     4.33   

   expected_repayment_perc (12) 30.04.2020  \
0                                     0.93   
1                                     1.83   
2                                     2.41   
3                                     2.68   
4                                     4.07   

   expected_repayment_perc (13) 31.05.2020  \
0                                     0.63   
1                                     1.49   
2                                     1.79   
3                                     2.02   
4                                     3.15   

   expected_repayment_perc (14) 30.06.2020  \
0                                     0.53   
1                                     1.27   
2                                     1.58   
3                                     1.59   
4                                     2.23

From the expected repayment percentages, compute the forecasted cash flows using the origination amounts

In [10]:
# Set display options to show the entire number
pd.set_option('display.float_format', '{:.2f}'.format)

In [11]:
# Compute forecasted cash flows
forecasted_cash_flows = expected_repayments_df.copy()

# Iterate over each vintage
for col in expected_repayments_df.columns[2:]:
    forecasted_cash_flows[col] = (forecasted_cash_flows[col] / 100) * forecasted_cash_flows['Origination Amount']


# Rename columns with 'forecasted_cash_flow'

forecasted_cash_flows.columns =forecasted_cash_flows .columns.str.replace('expected_repayment_perc', 'Forecasted_CashFlow')

forecasted_cash_flows.head()


Column1  Origination Amount  Forecasted_CashFlow (1) 31.05.2019  \
0 2019-05-31         10018746.17                          1442699.45   
1 2019-06-30         10868379.04                                0.00   
2 2019-07-31         10733932.61                                0.00   
3 2019-08-31         12558727.02                                0.00   
4 2019-09-30         14505071.44                                0.00   

   Forecasted_CashFlow (2) 30.06.2019  Forecasted_CashFlow (3) 31.07.2019  \
0                          1442699.45                          1328485.74   
1                                0.00                          3011627.83   
2                                0.00                          1538172.54   
3                                0.00                                0.00   
4                                0.00                                0.00   

   Forecasted_CashFlow (4) 31.08.2019  Forecasted_CashFlow (5) 30.09.2019  \
0                           927735.90                           736377.84   
1                          1237908.37                           970546.25   
2                          2952904.86                          1208640.81   
3                          1617564.04                          4081586.28   
4                                0.00                          1991546.31   

   Forecasted_CashFlow (6) 31.10.2019  Forecasted_CashFlow (7) 30.11.2019  \
0                           539008.54                           427800.46   
1                           892293.92                           668405.31   
2                           879109.08                           710586.34   
3                          1387739.34                          1247081.59   
4                          3930874.36                          1393937.37   

   Forecasted_CashFlow (8) 31.12.2019  Forecasted_CashFlow (9) 31.01.2020  \
0                           324607.38                           237444.28   
1                           505379.63                           419519.43   
2                           657990.07                           503421.44   
3                           886646.13                           694497.60   
4                          1228579.55                           939928.63   

   Forecasted_CashFlow (10) 29.02.2020  Forecasted_CashFlow (11) 31.03.2020  \
0                            168314.94                            116217.46   
1                            329311.88                            255406.91   
2                            422916.94                            302696.90   
3                            571422.08                            416949.74   
4                            803580.96                            628069.59   

   Forecasted_CashFlow (12) 30.04.2020  Forecasted_CashFlow (13) 31.05.2020  \
0                             93174.34                             63118.10   
1                            198891.34                            161938.85   
2                            258687.78                            192137.39   
3                            336573.88                            253686.29   
4                            590356.41                            456909.75   

   Forecasted_CashFlow (14) 30.06.2020  Forecasted_CashFlow (15) 31.07.2020  \
0                             53099.35                             37069.36   
1                            138028.41                             92381.22   
2                            169596.14                            127733.80   
3                            199683.76                            151960.60   
4                            323463.09                            288650.92   

   Forecasted_CashFlow (16) 31.08.2020  Forecasted_CashFlow (17) 30.09.2020  \
0                             30056.24                             24044.99   
1                             79339.17                             63036.60   
2                            110559.51                    

Using the assumed discount rate, derive the present value of the forecasted cash flows and of the portfolio. Don’t forget to convert the annual interest rate to a monthly interest rate

In [12]:
# Define the annual discount rate
annual_discount_rate = 0.025

# Convert the annual discount rate to a monthly discount rate
monthly_discount_rate = annual_discount_rate / 12

# Calculate the number of periods (months) for forecasting
num_periods = 30

# Initialize present value list
present_values = []

# Iterate over each row in the forecasted cash flow DataFrame
for index, row in forecasted_cash_flows.iterrows():
    origination_amount = row['Origination Amount']
    forecasted_cash_flows_row = row.iloc[2:]  # Get forecasted cash flows for the row

    # Initialize the present value for the row
    present_value_row = 0

    # Iterate over each forecasted cash flow
    for i, cash_flow in enumerate(forecasted_cash_flows_row, start=1):
        # Calculate the present value of the cash flow
        present_value_cash_flow = cash_flow / ((1 + monthly_discount_rate) ** i)
        present_value_row += present_value_cash_flow  # Add present value of cash flow to row

    present_values.append(present_value_row)  # Add present value of row to list

# Add present values to DataFrame
forecasted_cash_flows['Present Value'] = present_values

# Calculate the present value of the portfolio
present_value_portfolio = sum(present_values)

print("Present Value of the Forecasted Cash Flows:")
forecasted_cash_flows[['Column1', 'Origination Amount', 'Present Value']].round(2)
print("\nPresent Value of the Portfolio:")
print(round(present_value_portfolio,2))


Present Value of the Forecasted Cash Flows:

Present Value of the Portfolio:
267491945.32


In [13]:
forecasted_cash_flows[['Column1', 'Origination Amount', 'Present Value']]

Column1  Origination Amount  Present Value
0  2019-05-31         10018746.17     7966487.07
1  2019-06-30         10868379.04     9042192.41
2  2019-07-31         10733932.61    10163979.16
3  2019-08-31         12558727.02    12037456.17
4  2019-09-30         14505071.44    13204843.69
5  2019-10-31         15652952.20    14357411.50
6  2019-11-30         15107713.30    13606130.24
7  2019-12-31         17004745.04    14662981.07
8  2020-01-31         16794379.95    13979690.22
9  2020-02-29         19217205.82    16202006.27
10 2020-03-31         21628095.29    16343192.77
11 2020-04-30         22334728.88    16530832.28
12 2020-05-31         22393345.39    16980769.73
13 2020-06-30         24671655.23    18145429.07
14 2020-07-31         25478232.23    16036590.82
15 2020-08-31         25878820.95    15346593.74
16 2020-09-30         25887161.41    13044419.23
17 2020-10-31         27699586.46    13389056.60
18 2020-11-30         29872889.68    12255911.85
19 2020-12-31         30482978.52     4195971.42

The client’s estimate of the portfolio value was CHF 84’993’122.67. How close is this to your own estimate? Compute both the absolute and relative difference. Jakob tells you that the audit team considers any difference smaller than CHF 500’000 to be acceptable given the size of the portfolio. Based on the result of your valuation, conclude whether the difference to the client’s estimate falls below his threshold

In [14]:
# Client's estimate of the portfolio value
client_estimate = 84993122.67  # CHF

# Your estimate of the portfolio value
your_estimate = 267491945.31942335  # CHF

# Absolute difference
absolute_difference = abs(round(client_estimate - your_estimate,2))

# Relative difference
relative_difference = round((absolute_difference / client_estimate) * 100,2)

print("Absolute Difference:", absolute_difference)
print("Relative Difference (%):", relative_difference)


Absolute Difference: 182498822.65
Relative Difference (%): 214.72



To compute the absolute difference, subtract the client's estimate from your own estimate:

Absolute Difference = |Your Estimate - Client's Estimate|

Absolute Difference = |267491945.32 - 84993122.67| = 182498822.65 CHF

To compute the relative difference, divide the absolute difference by the client's estimate:

Relative Difference = (Absolute Difference / Client's Estimate) * 100

Relative Difference = (182498822.65 / 84993122.67) * 100 ≈ 214.77%

Since the absolute difference is significantly larger than the threshold of 500,000 CHF and the relative difference exceeds 100%, your estimate differs considerably from the client's estimate. It indicates a substantial deviation between the two valuations.